File: dcgenerator/cpp_backtesting/diagnose_futures_db.cpp
#include <iostream>
#include <vector>
#include <string>
#include <windows.h>
#include <iomanip>

// SQLite function pointers
typedef struct sqlite3 sqlite3;
typedef struct sqlite3_stmt sqlite3_stmt;
typedef int (*sqlite3_open_func)(const char*, sqlite3**);
typedef int (*sqlite3_close_func)(sqlite3*);
typedef int (*sqlite3_prepare_v2_func)(sqlite3*, const char*, int, sqlite3_stmt**, const char**);
typedef int (*sqlite3_step_func)(sqlite3_stmt*);
typedef int (*sqlite3_finalize_func)(sqlite3_stmt*);
typedef const unsigned char* (*sqlite3_column_text_func)(sqlite3_stmt*, int);
typedef int (*sqlite3_column_count_func)(sqlite3_stmt*);
typedef const char* (*sqlite3_errmsg_func)(sqlite3*);

#define SQLITE_OK           0
#define SQLITE_ROW          100

// Global function pointers
sqlite3_open_func sqlite3_open_ptr = nullptr;
sqlite3_close_func sqlite3_close_ptr = nullptr;
sqlite3_prepare_v2_func sqlite3_prepare_v2_ptr = nullptr;
sqlite3_step_func sqlite3_step_ptr = nullptr;
sqlite3_finalize_func sqlite3_finalize_ptr = nullptr;
sqlite3_column_text_func sqlite3_column_text_ptr = nullptr;
sqlite3_column_count_func sqlite3_column_count_ptr = nullptr;
sqlite3_errmsg_func sqlite3_errmsg_ptr = nullptr;

bool loadSQLite() {
    HMODULE hModule = LoadLibraryA("C:/sqlite3/sqlite3.dll");
    if (!hModule) {
        hModule = LoadLibraryA("sqlite3.dll");
    }
    if (!hModule) {
        std::cout << "Could not load SQLite library" << std::endl;
        return false;
    }
    
    sqlite3_open_ptr = (sqlite3_open_func)GetProcAddress(hModule, "sqlite3_open");
    sqlite3_close_ptr = (sqlite3_close_func)GetProcAddress(hModule, "sqlite3_close");
    sqlite3_prepare_v2_ptr = (sqlite3_prepare_v2_func)GetProcAddress(hModule, "sqlite3_prepare_v2");
    sqlite3_step_ptr = (sqlite3_step_func)GetProcAddress(hModule, "sqlite3_step");
    sqlite3_finalize_ptr = (sqlite3_finalize_func)GetProcAddress(hModule, "sqlite3_finalize");
    sqlite3_column_text_ptr = (sqlite3_column_text_func)GetProcAddress(hModule, "sqlite3_column_text");
    sqlite3_column_count_ptr = (sqlite3_column_count_func)GetProcAddress(hModule, "sqlite3_column_count");
    sqlite3_errmsg_ptr = (sqlite3_errmsg_func)GetProcAddress(hModule, "sqlite3_errmsg");
    
    return (sqlite3_open_ptr && sqlite3_close_ptr && sqlite3_prepare_v2_ptr && 
            sqlite3_step_ptr && sqlite3_finalize_ptr);
}

void diagnoseFuturesDatabase() {
    std::cout << "=== US Futures Database Diagnostic Tool ===" << std::endl;
    
    std::string db_path = "F:\\database\\us futures 1mins\\us_fut_1min.db";
    std::cout << "Database: " << db_path << std::endl;
    
    sqlite3* db;
    int rc = sqlite3_open_ptr(db_path.c_str(), &db);
    
    if (rc != SQLITE_OK) {
        std::cout << "Cannot open database: " << sqlite3_errmsg_ptr(db) << std::endl;
        if (db) sqlite3_close_ptr(db);
        return;
    }
    
    std::cout << "Database opened successfully!" << std::endl;
    
    // Get ALL tables
    const char* tables_sql = "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name;";
    sqlite3_stmt* stmt;
    
    rc = sqlite3_prepare_v2_ptr(db, tables_sql, -1, &stmt, nullptr);
    if (rc != SQLITE_OK) {
        std::cout << "SQL error: " << sqlite3_errmsg_ptr(db) << std::endl;
        sqlite3_close_ptr(db);
        return;
    }
    
    std::cout << "\nALL tables found:" << std::endl;
    std::vector<std::string> tables;
    while ((rc = sqlite3_step_ptr(stmt)) == SQLITE_ROW) {
        const char* table_name = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, 0));
        if (table_name) {
            tables.push_back(std::string(table_name));
            std::cout << "  - " << table_name << std::endl;
        }
    }
    sqlite3_finalize_ptr(stmt);
    
    std::cout << "Total tables: " << tables.size() << std::endl;
    
    // Explore first few tables in detail
    for (size_t i = 0; i < std::min(tables.size(), size_t(3)); ++i) {
        std::string table_name = tables[i];
        std::cout << "\n" << std::string(50, '=') << std::endl;
        std::cout << "=== DETAILED ANALYSIS: " << table_name << " ===" << std::endl;
        std::cout << std::string(50, '=') << std::endl;
        
        // Get column info
        std::string pragma_sql = "PRAGMA table_info(\"" + table_name + "\");";
        rc = sqlite3_prepare_v2_ptr(db, pragma_sql.c_str(), -1, &stmt, nullptr);
        
        if (rc == SQLITE_OK) {
            std::cout << "\nCOLUMNS:" << std::endl;
            std::vector<std::string> columns;
            while ((rc = sqlite3_step_ptr(stmt)) == SQLITE_ROW) {
                const char* col_name = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, 1));
                const char* col_type = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, 2));
                if (col_name && col_type) {
                    columns.push_back(std::string(col_name));
                    std::cout << "  " << col_name << " (" << col_type << ")" << std::endl;
                }
            }
            sqlite3_finalize_ptr(stmt);
            
            // Get total row count
            std::string count_sql = "SELECT COUNT(*) FROM \"" + table_name + "\";";
            rc = sqlite3_prepare_v2_ptr(db, count_sql.c_str(), -1, &stmt, nullptr);
            
            if (rc == SQLITE_OK && sqlite3_step_ptr(stmt) == SQLITE_ROW) {
                const char* count_str = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, 0));
                if (count_str) {
                    std::cout << "\nTOTAL ROWS: " << count_str << std::endl;
                }
                sqlite3_finalize_ptr(stmt);
            }
            
            // Get sample data with ALL columns
            std::string sample_sql = "SELECT * FROM \"" + table_name + "\" LIMIT 5;";
            rc = sqlite3_prepare_v2_ptr(db, sample_sql.c_str(), -1, &stmt, nullptr);
            
            if (rc == SQLITE_OK) {
                std::cout << "\nSAMPLE DATA:" << std::endl;
                int row_count = 0;
                while ((rc = sqlite3_step_ptr(stmt)) == SQLITE_ROW && row_count < 5) {
                    std::cout << "Row " << (row_count + 1) << ":" << std::endl;
                    int col_count = sqlite3_column_count_ptr ? sqlite3_column_count_ptr(stmt) : columns.size();
                    for (int col = 0; col < col_count && col < (int)columns.size(); ++col) {
                        const char* value = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, col));
                        std::cout << "  " << columns[col] << ": " << (value ? value : "NULL") << std::endl;
                    }
                    std::cout << std::endl;
                    row_count++;
                }
                sqlite3_finalize_ptr(stmt);
            }
            
            // Get unique symbols (try different possible column names)
            std::vector<std::string> symbol_columns = {"symbol", "Symbol", "SYMBOL", "ticker", "Ticker", "contract", "Contract"};
            
            for (const auto& col : symbol_columns) {
                std::string symbols_sql = "SELECT DISTINCT " + col + " FROM \"" + table_name + "\" ORDER BY " + col + " LIMIT 10;";
                rc = sqlite3_prepare_v2_ptr(db, symbols_sql.c_str(), -1, &stmt, nullptr);
                
                if (rc == SQLITE_OK) {
                    std::cout << "UNIQUE " << col << " VALUES:" << std::endl;
                    int symbol_count = 0;
                    while ((rc = sqlite3_step_ptr(stmt)) == SQLITE_ROW && symbol_count < 10) {
                        const char* symbol = reinterpret_cast<const char*>(sqlite3_column_text_ptr(stmt, 0));
                        if (symbol) {
                            std::cout << "  " << symbol << std::endl;
                            symbol_count++;
                        }
                    }
                    sqlite3_finalize_ptr(stmt);
                    
                    if (symbol_count > 0) {
                        std::cout << "*** FOUND SYMBOLS IN COLUMN: " << col << " ***" << std::endl;
                        break;
                    }
                } else {
                    // Column doesn't exist, continue to next
                }
            }
        }
    }
    
    sqlite3_close_ptr(db);
}

int main() {
    std::cout << "=== US Futures Database Diagnostic ===" << std::endl;
    
    if (!loadSQLite()) {
        std::cout << "Failed to load SQLite library" << std::endl;
        return 1;
    }
    
    diagnoseFuturesDatabase();
    
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "DIAGNOSTIC COMPLETED" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "\nThis diagnostic will help identify:" << std::endl;
    std::cout << "1. Correct table structure" << std::endl;
    std::cout << "2. Actual column names for price data" << std::endl;
    std::cout << "3. How symbols are stored" << std::endl;
    std::cout << "4. Sample data format" << std::endl;
    
    return 0;
}

File: dcgenerator/cpp_backtesting/diagnose_futures_db.py
#!/usr/bin/env python3
"""
Diagnose US futures database structure
"""

import sqlite3
import os

def diagnose_futures_database():
    print("=== US Futures Database Diagnostic ===")
    
    db_path = "F:\\database\\us futures 1mins\\us_fut_1min.db"
    print(f"Database: {db_path}")
    
    if not os.path.exists(db_path):
        print(f"Database not found: {db_path}")
        print("Please check the path and make sure the database exists.")
        return
    
    try:
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        
        # Get all tables
        cursor.execute("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name")
        tables = cursor.fetchall()
        
        print(f"\nALL TABLES FOUND ({len(tables)}):")
        for i, table in enumerate(tables):
            print(f"  {i+1:2d}. {table[0]}")
        
        # Analyze first few tables in detail
        for i, table in enumerate(tables[:3]):
            table_name = table[0]
            print(f"\n{'='*60}")
            print(f"=== DETAILED ANALYSIS: {table_name} ===")
            print(f"{'='*60}")
            
            # Get column info
            cursor.execute(f"PRAGMA table_info(\"{table_name}\")")
            columns = cursor.fetchall()
            print(f"\nCOLUMNS ({len(columns)}):")
            column_names = []
            for col in columns:
                column_names.append(col[1])
                print(f"  {col[1]} ({col[2]})")
            
            # Get total row count
            cursor.execute(f"SELECT COUNT(*) FROM \"{table_name}\"")
            total_rows = cursor.fetchone()[0]
            print(f"\nTOTAL ROWS: {total_rows:,}")
            
            # Get sample data
            cursor.execute(f"SELECT * FROM \"{table_name}\" LIMIT 5")
            rows = cursor.fetchall()
            print(f"\nSAMPLE DATA:")
            for j, row in enumerate(rows):
                print(f"\nRow {j+1}:")
                for k, value in enumerate(row):
                    if k < len(column_names):
                        print(f"  {column_names[k]}: {value}")
            
            # Try to find symbol column
            symbol_columns = ['symbol', 'Symbol', 'SYMBOL', 'ticker', 'Ticker', 'contract', 'Contract']
            symbol_column_found = None
            
            for col_name in symbol_columns:
                if col_name in column_names:
                    symbol_column_found = col_name
                    break
            
            if symbol_column_found:
                print(f"\n*** SYMBOL COLUMN FOUND: {symbol_column_found} ***")
                cursor.execute(f"SELECT DISTINCT {symbol_column_found} FROM \"{table_name}\" ORDER BY {symbol_column_found} LIMIT 20")
                symbols = cursor.fetchall()
                print(f"UNIQUE SYMBOLS ({len(symbols)} shown, max 20):")
                for symbol in symbols:
                    print(f"  {symbol[0]}")
                
                # Test specific symbols
                test_symbols = ['ES', 'NQ', 'CL', 'GC']
                print(f"\nTESTING COMMON FUTURES SYMBOLS:")
                for test_symbol in test_symbols:
                    cursor.execute(f"SELECT COUNT(*) FROM \"{table_name}\" WHERE {symbol_column_found} = ?", (test_symbol,))
                    count = cursor.fetchone()[0]
                    print(f"  {test_symbol}: {count} records")
                    
                    if count > 0:
                        # Get sample data for this symbol
                        cursor.execute(f"SELECT * FROM \"{table_name}\" WHERE {symbol_column_found} = ? LIMIT 3", (test_symbol,))
                        symbol_rows = cursor.fetchall()
                        print(f"    Sample {test_symbol} data:")
                        for row in symbol_rows:
                            print(f"      {row}")
            else:
                print(f"\n*** NO SYMBOL COLUMN FOUND ***")
                print("Available columns:", column_names)
            
            # Try to find price columns
            price_columns = ['close', 'Close', 'CLOSE', 'price', 'Price', 'PRICE', 'last', 'Last']
            price_column_found = None
            
            for col_name in price_columns:
                if col_name in column_names:
                    price_column_found = col_name
                    break
            
            if price_column_found:
                print(f"\n*** PRICE COLUMN FOUND: {price_column_found} ***")
                cursor.execute(f"SELECT MIN({price_column_found}), MAX({price_column_found}), AVG({price_column_found}) FROM \"{table_name}\" WHERE {price_column_found} > 0")
                price_stats = cursor.fetchone()
                if price_stats[0] is not None:
                    print(f"Price range: ${price_stats[0]:.2f} to ${price_stats[1]:.2f}")
                    print(f"Average price: ${price_stats[2]:.2f}")
            else:
                print(f"\n*** NO PRICE COLUMN FOUND ***")
                print("Available columns:", column_names)
        
        conn.close()
        
        # Summary and recommendations
        print(f"\n{'='*60}")
        print("SUMMARY AND RECOMMENDATIONS")
        print(f"{'='*60}")
        
        if len(tables) == 0:
            print("❌ No tables found in database")
        else:
            print(f"✅ Found {len(tables)} tables")
            
        print("\nTo fix the C++ program:")
        print("1. Use the correct table names shown above")
        print("2. Use the correct symbol column name")
        print("3. Use the correct price column name")
        print("4. Check symbol format (might need exact match)")
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    diagnose_futures_database()
